File: require/vigenere-cipher.rb
ALPHABET = %w[a b c d e f g h i j k l m n o p q r s t u v w x y z]

class VigenereCipher

  def initialize(key, alphabet = ALPHABET)
    @key, @alphabet = key.downcase, alphabet
  end

  def cipher_message(message)
    i = 0    # Iterates over key
    ciphered_msg = ""
    message.gsub(/[ |\n|\t]/, '')
      .downcase
      .each_char do |c|
        ciphered_msg << cipher_char(c, i % @key.size)
        i += 1
      end
    ciphered_msg
  end

  def cipher_char(c, index)
    @alphabet[(@alphabet.index(c) + @alphabet.index(@key[index])) % @alphabet.size]
  end

  
  def decipher_message(message)
    i = 0    # Iterates over key
    deciphered_msg = ""
    message.gsub(/[ |\n|\t]/, '')
    .downcase
    .each_char do |c|
      deciphered_msg << decipher_char(c, i % @key.size)
      i += 1
    end
    deciphered_msg
  end

  def decipher_char(c, index)
    @alphabet[(@alphabet.index(c) - @alphabet.index(@key[index])) % @alphabet.size]
  end


end

File: web/scripts/vigenere-cipher.js



export function VigenereCipher(key = "mision", alphabet = []) {
    if (alphabet.length === 0) {
      alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
        'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 
        's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
    }

    this.key = key;
    this.alphabet = alphabet;

    this.cipherChar = function(char, index) {
      return this.charCipher(char, index);
    };

    this.decipherChar = function(char, index) {
      return this.charCipher(char, index, false);
    };

    this.cipherMessage = function(message) {
      return this.messageCipher(message)  ;
    };

    this.decipherMessage = function(message) {
      return this.messageCipher(message, false);
    };


    this.charCipher = function(char, index, cipher = true) {
      const key = this.key
      const alphabet = this.alphabet
      
      if (index < 0 || index >= key.length) {
        return undefined;
      }

      if (alphabet.indexOf(char) < 0) {
        return undefined;
      }
      
      const charIndex = alphabet.indexOf(char);
      const keyCharIndex = alphabet.indexOf(key[index]);
      
      const offset = charIndex + (cipher ? keyCharIndex : -keyCharIndex);
      const cipheredCharIndex = (offset < 0 ? alphabet.length + offset : offset) % alphabet.length;
      const result = alphabet[cipheredCharIndex];

      return result;
    };

    this.messageCipher = function(message, cipher = true) {

      let keyIndex = 0;
      let cipheredMessage = "";

      message.split('').map( letter => {
        let cipheredChar = cipher ?  this.cipherChar(letter, keyIndex % this.key.length) : this.decipherChar(letter, keyIndex % this.key.length);
        if (cipheredChar) {
          cipheredMessage += cipheredChar;
          keyIndex++;
        }
      });

      return cipheredMessage;
    };
};

 



